src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+const SERVICE_ID = 'DisplayObserver';
+
+const DOC_OBSERVER = 0;
+const VIEWPORT_OBSERVER = 1;
+
+/** @implements {Disposable} */
+export class DisplayObserver {
+  /**
+   * @param {!AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.docObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, DOC_OBSERVER),
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
+      }
+    );
+
+    /** @private @const {!IntersectionObserver} */
+    this.viewportObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, VIEWPORT_OBSERVER),
+      {threshold: 0.001}
+    );
+
+    /** @private {boolean} */
+    this.isDocVisible_ = ampdoc.isVisible();
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() => {
+      if (ampdoc.isVisible() !== this.isDocVisible_) {
+        this.isDocVisible_ = ampdoc.isVisible();
+        this.docVisibilityChanged_();
+      }
+    });
+
+    /** @private @const {!Map<!Element, !Array<function(boolean)>>} */
+    this.targetObserverMultimap_ = new Map();
+
+    /** @private @const {!Map<!Element, !Array<boolean>>} */
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverMultimap_.set(target, callbacks);
+      this.docObserver_.observe(target);
+      this.viewportObserver_.observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      if (this.targetDisplayMap_.has(target)) {
```

Comment:
Can you add a comment here? I think what's happening is that if we already observed this element, and it's already had observers fire for it, then we're queueing a task to notify this new callback of the already recorded state?

Reply:
Correct. And I added the comment.